code/modules/urist/modules/GOAI/integrations/base_commander/mob_commander/obstacle_handlers/airlocks.dm

```diff
@@ -31,260 +39,565 @@
+
+		if(result in to_find)
+			OBSTACLE_DEBUG_LOG("Wanted result for [airlock]")
+			if(cut)
+				wire_queue[picked_wire] = "cut_wire"
+			to_find -= result
+			tracker.BBSet("TargetWires", to_find)
 		else
-			//Todo: pulse stuff here
+			OBSTACLE_DEBUG_LOG("Unwanted result for [airlock]")
+			wire_queue[picked_wire] = "cut_and_mend"
+			tried_wires.Add(picked_wire)
+			tracker.BBSet("TriedWires", tried_wires)
+
+		tracker.BBSet("LastAction", world.time)
+
+/datum/goai/mob_commander/proc/HandleAirlockAI()
+	//todo
+	return
+
+/datum/goai/mob_commander/proc/HandleAirlockPanelScrew(var/datum/ActionTracker/tracker, var/obj/machinery/door/airlock/airlock, var/open)
+	var/mob/living/mob_pawn = src.GetPawn()
+	if(!istype(mob_pawn) || isnull(open))
+		tracker.SetFailed()
+		return
+
+	airlock = resolve_weakref(airlock)
+
+	if(!istype(airlock))
+		tracker.SetFailed()
+		return
+
+	if(TimedOutWalkDist(tracker, mob_pawn, airlock, 15 SECONDS))
+		DropObstacleMemory(airlock)
+		OBSTACLE_DEBUG_LOG("Tracker timed-out for [airlock] [COORDS_TUPLE(airlock)] panel screw task")
+		return
+
+	if(tracker.BBGet("InProgress", FALSE))
+		return
 
-	if(!length(wires))
+	if(airlock in src.brain.perceptions[SENSE_SIGHT_CURR] && airlock.p_open == open)
 		tracker.SetDone()
 		return
 
-	//var/list/wireColours = list("red", "blue", "green", "darkred", "orange", "brown", "gold", "gray", "cyan", "navy", "purple", "pink", "black", "yellow")
-	/*
-	var/const/AIRLOCK_WIRE_IDSCAN = 1
-	var/const/AIRLOCK_WIRE_MAIN_POWER1 = 2
-	var/const/AIRLOCK_WIRE_MAIN_POWER2 = 4
-	var/const/AIRLOCK_WIRE_DOOR_BOLTS = 8
-	var/const/AIRLOCK_WIRE_BACKUP_POWER1 = 16
-	var/const/AIRLOCK_WIRE_BACKUP_POWER2 = 32
-	var/const/AIRLOCK_WIRE_OPEN_DOOR = 64
-	var/const/AIRLOCK_WIRE_AI_CONTROL = 128
-	var/const/AIRLOCK_WIRE_ELECTRIFY = 256
-	var/const/AIRLOCK_WIRE_SAFETY = 512
-	var/const/AIRLOCK_WIRE_SPEED = 1024
-	var/const/AIRLOCK_WIRE_LIGHT = 2048
-	*/
\ No newline at end of file
+	if(!NavigateNextTo(tracker, mob_pawn, airlock))
+		return
+
+	tracker.BBSet("InProgress", TRUE)
+
+	if(UseGenericTool(mob_pawn, airlock, "isscrewdriver", STATE_HASSCREWDRIVER))
+		tracker.SetDone()
+	else
+		tracker.SetFailed()
+		DropObstacleMemory(airlock)
+
+/datum/goai/mob_commander/proc/HandleAirlockPryOpen(var/datum/ActionTracker/tracker, var/obj/machinery/door/airlock/airlock, var/open)
+	var/mob/living/mob_pawn = src.GetPawn()
+	if(!istype(mob_pawn) || isnull(open))
+		tracker.SetFailed()
+		return
+
+	airlock = resolve_weakref(airlock)
+
+	if(isnull(airlock))
+		tracker.SetDone()
+		return
+	else if(!istype(airlock))
+		tracker.SetFailed()
+		return
+
+	if(TimedOutWalkDist(tracker, mob_pawn, airlock, 15 SECONDS))
+		DropObstacleMemory(airlock)
+		OBSTACLE_DEBUG_LOG("Tracker timed-out for [airlock] [COORDS_TUPLE(airlock)] pry open task")
+		return
+
+	if(tracker.BBGet("InProgress", FALSE))
+		return
+
+	if(airlock in src.brain.perceptions[SENSE_SIGHT_CURR] && airlock.density != open)
```

Comment:
Unsafe, didn't account for null refs here!

Reply:
There's a null check on L573 so this should be safe?

Reply:
That's not what I'm referring to. We've checked that airlock is not null - that's fine.

What we don't know at this point is whether src has a Brain, and if it does - whether that Brain has Perceptions initialized.

If Brain is nulled (e.g. we tried to delete the AI, but the handler is still running), this will runtime out and go crazy. I'd need to dig through the code to remember if Perceptions is lazy-initialized, but potentially both of those derefs can be null.

Reply:
Ah, my bad. Totally didn't think of the brain being null here.